src/stringutils.cpp

```diff
@@ -152,4 +152,36 @@ std::optional<int64_t> stoi64(const std::string& str) {
         return std::nullopt;
     }
 }
+
+bool isValidUtf8(const std::string& text) {
+    // inspect the chars from the end of the string to test if the utf8 sequence is complete
+    int byte_counter = 0;
+    for (int i = text.size() - 1; i >= 0 && byte_counter <= 3; i--) {
+        int x = static_cast<int>(static_cast<unsigned char>(text[i]));
+        if ((x >> 7) == 0b0)
+            return true;         // last char is a single byte char
+        if ((x >> 6) == 0b10) {  // octet belong to multibyte sequence
+            byte_counter++;
+        } else if ((x >> 5) == 0b110) {  // first byte of 2 byte sequence
+            if (byte_counter + 1 == 2)
+                return true;
+            else
+                return false;
+        } else if ((x >> 4) == 0b1110) {  // first byte of 3 byte sequence
+            if (byte_counter + 1 == 3)
+                return true;
+            else
+                return false;
+        } else if ((x >> 3) == 0b11110) {  // first byte of 3 byte sequence
+            if (byte_counter + 1 == 4)
+                return true;
+            else
+                return false;
```

Comment:
We could make this block shorter if conditionals are refactored to for example:
```
else if ((x >> <shift> == <val>) && (byte_counter + 1 != <val>))
	return false;
```
Then conditions only check bad values and if it's not detected we get to the end of that function so the `return true` at the end is effective.